src/groups/curves/short_weierstrass/mod.rs

```diff
@@ -221,6 +221,43 @@ where
 
         Ok(Self::new(x, y, z))
     }
+
+    /// Mixed addition, which is useful when `other` is known to have z = 1.
+    #[tracing::instrument(target = "r1cs", skip(self, other_x, other_y))]
+    pub(crate) fn add_mixed(&self, (other_x, other_y): (&F, &F)) -> Result<Self, SynthesisError> {
+        // Complete mixed addition formula from Renes-Costello-Batina 2015
+        // Algorithm 2
+        // (https://eprint.iacr.org/2015/1060).
+        //
+        // Adapted from code in
+        // https://github.com/RustCrypto/elliptic-curves/blob/master/p256/src/arithmetic.rs
+        let three_b = P::COEFF_B.double() + &P::COEFF_B;
+
+        let xx = &self.x * other_x; // 1
+        let yy = &self.y * other_y; // 2
+        let xy_pairs = ((&self.x + &self.y) * &(other_x + other_y)) - (&xx + &yy); // 4, 5, 6, 7, 8
+        let xz_pairs = (other_x * &self.z) + &self.x; // 8, 9
+        let yz_pairs = (other_y * &self.z) + &self.y; // 10, 11
+
+        let axz = mul_by_coeff_a::<P, F>(&xz_pairs); // 12
+
+        let bz3_part = &axz + &self.z * three_b; // 13, 14
+
+        let yy_m_bz3 = &yy - &bz3_part; // 15
+        let yy_p_bz3 = &yy + &bz3_part; // 16
+
+        let azz = mul_by_coeff_a::<P, F>(&self.z); // 20
+        let xx3_p_azz = xx.double().unwrap() + &xx + &azz; // 18, 19, 22
+
+        let bxz3 = &xz_pairs * three_b; // 21
+        let b3_xz_pairs = mul_by_coeff_a::<P, F>(&(&xx - &azz)) + &bxz3; // 23, 24, 25
+
+        let x = (&yy_m_bz3 * &xy_pairs) - &yz_pairs * &b3_xz_pairs; // 28,29, 30
+        let y = (&yy_p_bz3 * &yy_m_bz3) + &xx3_p_azz * b3_xz_pairs; // 17, 26, 27
+        let z = (&yy_p_bz3 * &yz_pairs) + xy_pairs * xx3_p_azz; // 31, 32, 33
```

Comment:
I think the following style of comments would help improve readability / checkability?
```suggestion
        // 1, 2     
        // xx = x1 * x2
        // yy = y1 * y2
        let xx = &self.x * other_x;
        let yy = &self.y * other_y;
        // 4, 5, 6, 7, 8
        // xy_pairs = (x1 + y1) * (x2 + y2) - (xx + yy)
        let xy_pairs = ((&self.x + &self.y) * &(other_x + other_y)) - (&xx + &yy);
        // 8,9,10,11
        // xz_pairs = (x2 * z1) + x1
        // yz_pairs = (y2 * z1) + y1
        let xz_pairs = (other_x * &self.z) + &self.x;
        let yz_pairs = (other_y * &self.z) + &self.y;
...
```

Reply:
if the only change is to rename self.x and other.x to x1 and x2, then I can make that change in code directly.

Reply:
oh cool, lets do that then

Reply:
Done